### ISSUE TITLE
Impersonate - User list not correct

### ISSUE THREAD

**Issue by GerritRega** [rajyraman/Levelup-for-Dynamics-CRM#216]
Browser Version: 5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.5112.102 Safari/537.36 Edg/104.0.1293.70
    Extension Version: 3.5.8
    ----------------------------------------------------------
it's the same pn Crome with 3.5.9

I try to use impersonate, but the list is cutted at the end.
I'm not able to impersonate, because the user is not listed

**Comment by enzoporcasi** [comment]
Hi, I have the same problem. Do you solved?

**Comment by rajyraman** [comment]
@GerritRega @enzoporcasi - Pushed out new release over the weekend. Please check with v3.6.0

**Comment by enzoporcasi** [comment]
Hi,
after the update of extension the list of user is empty.

regards
Enzo
[image: image.png]

Il giorno lun 10 ott 2022 alle ore 01:10 Natraj Yegnaraman <
***@***.***> ha scritto:

**Comment by rajyraman** [comment]
@enzoporcasi - Can you run the below FetchXML using FetchXML Builder and see if you are getting any results. This is the query which is used for showing the users. I checked in my environment, and I am able to see the list of users to impersonate.

```xml
<fetch version='1.0' output-format='xml-platform' mapping='logical' distinct='true'>
  <entity name='systemuser'>
    <attribute name='fullname' />
    <attribute name='azureactivedirectoryobjectid' />
    <filter>
      <condition attribute='isdisabled' operator='eq' value='0' />
      <condition attribute='islicensed' operator='eq' value='1' />
      <condition attribute='systemuserid' operator='eq-userid' />
      <condition attribute='accessmode' operator='eq' value='0' />
    </filter>
    <order attribute='fullname' descending='false' />
  </entity>
</fetch>
```

**Comment by GerritRega** [comment]
My list is also empty.
With your fetch XML i get my own user.

If i remove this line: <condition attribute='systemuserid' operator='ne-userid' />
i get 5000 users, and with SQL4CDS more then 5000 Users.

**Comment by rajyraman** [comment]
@GerritRega - I update the FetchXML couple of seconds after I posted the comment. Not sure if you are using the latest one. You should be getting all licensed users except your user account. Can you please run the query on my previous comment?

**Comment by GerritRega** [comment]
@rajyraman: Yes the user would be displayed now. FetchXML tester only show 5000 users.

**Comment by rajyraman** [comment]
@GerritRega - Are you saying it is showing the users now?

**Comment by GerritRega** [comment]
@rajyraman  yes, but not all. I'm not sure if it's a restriction of FetchXML tester. It stops at 5000

**Comment by rajyraman** [comment]
@GerritRega - I experienced this when I clicked the extension during page load. I clicked again to close the popup and clicked one more time, and all the users are visible.

**Comment by GerritRega** [comment]
@rajyraman i don't get the user inside the list, even the page was loaded

**Comment by C485** [comment]
Hello,

It's still broken, list is empty, i've checked api response and there is list of 4999 users.
Api call has status 200 so it's ok. There is no errors in console.
![image](https://user-images.githubusercontent.com/20170118/195042340-914b3819-4dd6-4c8b-8a44-49178e98ed73.png)

**Comment by rajyraman** [comment]
@C485 - So, you just have only environment open in 1 tab. Maybe we can do a screenshare. Can you send an email to natraj@dreamingincrm.com? I cannot still repro this error.

**Comment by C485** [comment]
> @C485 - So, you just have only environment open in 1 tab. Maybe we can do a screenshare. Can you send an email to [natraj@dreamingincrm.com](mailto:natraj@dreamingincrm.com)? I cannot still repro this error.

I'll check it by my self after work hours, i cannot share you the screen because im working with clients data.
Is 'master' branch te newest one? I'll compile this extension by my self and test everything.

**Comment by rajyraman** [comment]
@C485 - No worries. It is users-paging branch which has the latest changes. Would be great if you look into it and submit PR. Thank you.

**Comment by rajyraman** [comment]
@C485 - You also need Docker Desktop and Dev Containers extension to compile the code.

**Comment by GerritRega** [comment]
@rajyraman maybe this additional info could help you:
We have a productive environment and an test environment.
Inside the Productive system it's not working, inside the test environment it's working.

If i started it inside the testsystem and open the productive system i get users.

**Comment by rajyraman** [comment]
@GerritRega - I tried this today morning, both in Chrome and Edge. 2 tabs with 2 different environments normal mode - works. 2 tabs with 2 different environments, one is normal tab and one in incognito tab. That works as well. I will have to figure a way to repro this issue. Can you see the system users being retrieved in the devtools traffic?

**Comment by C485** [comment]
Sorry, had no time yesterday, today i'll have no time as well, here are some details:

there is silent exception:
![image](https://user-images.githubusercontent.com/20170118/195279877-06c99cb1-f5dc-43c0-a2d4-a79c50df920f.png)
![image](https://user-images.githubusercontent.com/20170118/195279944-0b2e25c3-7188-48b9-aec8-dd02906f01a0.png)

exactly at this line:
![image](https://user-images.githubusercontent.com/20170118/195280309-3d9d2a41-5327-446c-88d6-85a2fbd04bec.png)

where 'i' stands for:
![image](https://user-images.githubusercontent.com/20170118/195280436-70c308ac-eb2f-4ddc-8881-f620218a412f.png)


sooo, error comes from from empty pagecookie from response of 1 page, that's quite interesting


response from api looks like this:
![image](https://user-images.githubusercontent.com/20170118/195280869-f2ea8892-6ba3-4a1b-b61e-ae5c8bc1e1a2.png)

mby thats befause you're using v8.0 version of api?
home that will help you

**Comment by C485** [comment]
@rajyraman it's not all, i fixed this in debugger and extensions fails again here:
![image](https://user-images.githubusercontent.com/20170118/195290726-545252b8-16c9-48a2-a215-9f58f46ff189.png)
there is 31k users :D and i quess it's too much 
![image](https://user-images.githubusercontent.com/20170118/195290893-d32985b0-2c45-4810-a49e-043b4b714b29.png)

**Comment by C485** [comment]
stacktrace:
![image](https://user-images.githubusercontent.com/20170118/195291961-fdef2531-8932-4948-9a4d-f5a7169c3168.png)

**Comment by GerritRega** [comment]
Hello @rajyraman and @C485 Do you have an update? Are somebody able to see the issue ?

**Comment by C485** [comment]
> Hello @rajyraman and @C485 Do you have an update? Are somebody able to see the issue ?

Hi, i found the issue, but sadly i had no free time to make a fix. There're actually two problems and they need to be fixed to make it work.

**Comment by rajyraman** [comment]
@C485 - I am still stuck for time. I had a look at the issue on the date when the screen shot was posted, but I need to setup the environment or test data+fiddler so that post message crashes. It seems this happens only if you have lot of users. I have under 500 users only.

**Comment by GerritRega** [comment]
@rajyraman @C485  Thanks for your really quick response. I'll wait. We have round about 5500 active Users and miss this impersionate :-D

**Comment by GerritRega** [comment]
Hi @rajyraman  and @C485 Do you have an update here?

**Comment by rajyraman** [comment]
In the latest release v3.6.2 I have updated the extension to V3 manifest and removed the dropdown list of users due to the fact that there are people with more than 10000 users.

What you will now see is just a textbox where you enter the username of the user to be impersonated e.g. user@org.crm.onmicrosoft.com. There is no feedback if that username does not exist or if it is unlicensed because I had to rush this release due to a bugfix. But if the username is correct clicking the impersonate toggle will reload the page and load the app as the impersonated user.

I have also added the reset button on the impersonate tab to do a reset the impersonation and get back to the original user session because there were so many tickets about getting stuck with the impersonated session.
